File: python/src/bursar/credits/postgres/repositories/catalog.py
from __future__ import annotations

import json

from pydantic import ValidationError

from bursar.credits.postgres.repositories._types import DbQuery
from bursar.credits.postgres.repositories._utils import optional_mapping_row, require_mapping_row
from bursar.credits.postgres.repositories.schemas import CatalogRevisionRow, CatalogRevisionSummaryRow
from bursar.errors import StoreError


class CatalogRepository:
    def __init__(self, callproc: DbQuery) -> None:
        self._callproc = callproc

    @staticmethod
    def _parse_revision(row: dict[str, object]) -> CatalogRevisionRow:
        parsed = {
            "id": str(row.get("id")),
            "config": row.get("source_document"),
            "version": row.get("revision_no"),
            "label": row.get("label"),
            "active": row.get("status") == "active",
            "status": row.get("status"),
            "created_at": row.get("created_at"),
        }
        try:
            return CatalogRevisionRow.model_validate(parsed)
        except ValidationError as error:
            raise StoreError("CatalogRepository returned an invalid catalog revision") from error

    def get_active_catalog(self) -> CatalogRevisionRow | None:
        rows = self._callproc("active_catalog_revision", []) or []
        row = optional_mapping_row(rows, "CatalogRepository.get_active_catalog")
        if row is None or all(value is None for value in row.values()):
            return None
        return self._parse_revision(row)

    def publish_and_activate_catalog(
        self,
        config: str,
        label: str | None,
        rollout: dict[str, object],
    ) -> CatalogRevisionRow:
        return self._publish_revision(config, label, rollout, activate=True)

    def publish_catalog_draft(self, config: str, label: str | None) -> CatalogRevisionRow:
        return self._publish_revision(config, label, {"plans": {}}, activate=False)

    def _publish_revision(
        self,
        config: str,
        label: str | None,
        rollout: dict[str, object],
        *,
        activate: bool,
    ) -> CatalogRevisionRow:
        row = require_mapping_row(
            self._callproc(
                "publish_and_activate_catalog",
                [1, json.loads(config), label, activate, rollout],
            ),
            "CatalogRepository.publish_revision",
        )
        revision_no = row.get("revision_no")
        if not isinstance(revision_no, int):
            raise StoreError("CatalogRepository.publish_revision returned an invalid revision_no")
        revision = self.get_catalog_revision(revision_no)
        if revision is None:
            raise StoreError(f"Catalog revision {revision_no} disappeared after publication")
        return revision

    def get_catalog_history(self) -> list[CatalogRevisionSummaryRow]:
        rows = self._callproc("list_catalog_revisions", [500]) or []
        revisions: list[CatalogRevisionSummaryRow] = []
        for row in rows:
            if not isinstance(row, dict):
                raise StoreError("CatalogRepository.get_catalog_history returned a non-object row")
            revision = self._parse_revision(row)
            revisions.append(
                CatalogRevisionSummaryRow(
                    id=revision.id,
                    version=revision.version,
                    label=revision.label,
                    active=revision.active,
                    created_at=revision.created_at,
                )
            )
        return revisions

    def get_catalog_revision(self, version: int) -> CatalogRevisionRow | None:
        rows = self._callproc("catalog_revision_by_number", [version]) or []
        row = optional_mapping_row(rows, "CatalogRepository.get_catalog_revision")
        if row is None or all(value is None for value in row.values()):
            return None
        return self._parse_revision(row)

    def activate_catalog_revision(
        self,
        version: int,
        rollout: dict[str, object],
    ) -> CatalogRevisionRow:
        rows = self._callproc("activate_catalog_revision", [version, rollout]) or []
        row = optional_mapping_row(rows, "CatalogRepository.activate_catalog_revision")
        if row is None or all(value is None for value in row.values()):
            raise StoreError(f"Catalog revision {version} was not found")
        return self._parse_revision(row)

File: javascript/src/credits/postgres/repositories/catalog.ts
import { z } from "zod";
import { StoreError } from "../../../errors.js";
import type { CallProc } from "../../../shared/postgres-types.js";
import {
  optionalRecordRow,
  pgBoolean,
  requireRecordRow,
  safeParse,
} from "../../../shared/postgres-validation.js";

const CatalogRevisionRowSchema = z
  .object({
    id: z.string().min(1),
    config: z.record(z.string(), z.unknown()),
    version: z.coerce.number().int().positive(),
    label: z.string().nullable(),
    active: pgBoolean,
    status: z.enum(["draft", "published", "active", "retired"]),
    created_at: z.union([z.string(), z.date().transform((v) => v.toISOString())]),
  })
  .strict();
const PublishedRevisionSchema = z
  .object({
    revision_id: z.string().min(1),
    revision_no: z.coerce.number().int().positive(),
    status: z.enum(["published", "active"]),
  })
  .strict();
export type CatalogRevisionRow = z.infer<typeof CatalogRevisionRowSchema>;

export class CatalogRepository {
  constructor(private readonly callproc: CallProc) {}

  private parseRevision(row: Record<string, unknown>, context: string): CatalogRevisionRow {
    return safeParse(
      CatalogRevisionRowSchema,
      {
        id: row.id,
        config: row.source_document,
        version: row.revision_no,
        label: row.label,
        active: row.status === "active",
        status: row.status,
        created_at: row.created_at,
      },
      context,
    );
  }

  async getActiveCatalog(): Promise<CatalogRevisionRow | null> {
    const rows = await this.callproc("active_catalog_revision", []);
    const row = optionalRecordRow(rows, "CatalogRepository.getActiveCatalog");
    if (row === null || Object.values(row).every((value) => value === null)) return null;
    return this.parseRevision(row, "CatalogRepository.getActiveCatalog");
  }

  async publishAndActivateCatalog(
    config: string,
    label: string | null,
    rollout: Record<string, unknown>,
  ): Promise<CatalogRevisionRow> {
    return this.publishRevision(config, label, true, rollout);
  }

  async publishCatalogDraft(config: string, label: string | null): Promise<CatalogRevisionRow> {
    return this.publishRevision(config, label, false, { plans: {} });
  }

  private async publishRevision(
    config: string,
    label: string | null,
    activate = false,
    rollout: Record<string, unknown> = { plans: {} },
  ): Promise<CatalogRevisionRow> {
    const published = safeParse(
      PublishedRevisionSchema,
      requireRecordRow(
        await this.callproc("publish_and_activate_catalog", [
          1,
          JSON.parse(config),
          label,
          activate,
          rollout,
        ]),
        "CatalogRepository.publishRevision",
      ),
      "CatalogRepository.publishRevision",
    );
    const revision = await this.getCatalogRevision(published.revision_no);
    if (!revision) {
      throw new StoreError(
        `Catalog revision ${published.revision_no} disappeared after publication`,
      );
    }
    return revision;
  }

  async getCatalogHistory(): Promise<CatalogRevisionRow[]> {
    const rows = await this.callproc("list_catalog_revisions", [500]);
    return (rows ?? []).map((row) => {
      if (typeof row !== "object" || row === null || Array.isArray(row)) {
        throw new StoreError("CatalogRepository.getCatalogHistory returned a non-object row");
      }
      return this.parseRevision(
        row as Record<string, unknown>,
        "CatalogRepository.getCatalogHistory",
      );
    });
  }

  async getCatalogRevision(version: number): Promise<CatalogRevisionRow | null> {
    const rows = await this.callproc("catalog_revision_by_number", [version]);
    const row = optionalRecordRow(rows, "CatalogRepository.getCatalogRevision");
    if (row === null || Object.values(row).every((value) => value === null)) return null;
    return this.parseRevision(row, "CatalogRepository.getCatalogRevision");
  }

  async activateCatalogRevision(
    version: number,
    rollout: Record<string, unknown>,
  ): Promise<CatalogRevisionRow> {
    const rows = await this.callproc("activate_catalog_revision", [version, rollout]);
    const row = optionalRecordRow(rows, "CatalogRepository.activateCatalogRevision");
    if (row === null || Object.values(row).every((value) => value === null)) {
      throw new StoreError(`Catalog revision ${version} was not found`);
    }
    return this.parseRevision(row, "CatalogRepository.activateCatalogRevision");
  }
}
